File: apps/kinh-dich-service/src/application/usecases.ts
/**
 * Kinh Dich Service — Application Use Cases
 *
 * Orchestrates reading_composer module + repository ports.
 * No direct I/O — only ports.
 *
 * P2-KD-F: Rewired from domain/services.ts computeReading() to
 * reading_composer composeReading() per G5a handoff.
 * QUE_META sourced from domain/hexagram-data.ts (thin data module).
 */

import { composeReading, type ReadingComposerDependencies } from '../module/reading_composer/index.js';
import { QUE_META } from '../domain/hexagram-data.js';
import type { KinhDichRepositoryPort, PriceScorePort } from '../domain/repositories.js';
import type { ReadingRequest, ReadingResponse, MarketReadingResponse } from './dtos.js';
import { InsufficientDataError, HexagramNotFoundError } from '../domain/errors.js';

export class ReadingUseCase {
  constructor(
    private readonly repo: KinhDichRepositoryPort,
    private readonly priceScorePort: PriceScorePort,
    private readonly composerDeps: ReadingComposerDependencies,
  ) {}

  async execute(req: ReadingRequest): Promise<ReadingResponse> {
    const { stockCode, days = 30 } = req;
    const code = stockCode.toUpperCase();

    // Try to compute live reading from price scores first
    const scores = this.priceScorePort.computeScores(code, days);

    if (scores) {
      const reading = await composeReading(code, scores, this.composerDeps);
      return {
        stock: code,
        hexagram: reading.queChiNh.number,
        name: reading.queChiNh.name,
        trend: reading.queChiNh.trend,
        signal: reading.queChiNh.tradingSignal,
        confidence: reading.queChiNh.confidence,
        actionNote: reading.actionNote,
        overallReading: reading.overallReading,
        timestamp: reading.timestamp,
      };
    }

    // Fall back to stored reading
    const stored = this.repo.getLatestReading(code);
    if (!stored) {
      throw new HexagramNotFoundError(code);
    }

    // Build a minimal response from stored reading.
    // Use QUE_META to resolve the correct name for stored.hexagram_number — do NOT
    // compute a fresh reading with placeholder scores because the placeholder scores
    // always resolve to the same hexagram (not the stored one), causing the wrong name
    // to appear for every stock.
    const storedMeta = QUE_META.find((q) => q.id === stored.hexagram_number);
    const storedName = storedMeta?.name ?? `Que ${stored.hexagram_number}`;

    const fallbackScores = [0.1, 0.1, 0.1, 0.1, 0.1, 0.1]; // neutral placeholder
    const reading = await composeReading(code, fallbackScores, this.composerDeps);

    return {
      stock: code,
      hexagram: stored.hexagram_number,
      name: storedName,
      trend: reading.queChiNh.trend,
      signal: stored.trading_signal ?? reading.queChiNh.tradingSignal,
      confidence: stored.confidence ?? reading.queChiNh.confidence,
      actionNote: reading.actionNote,
      overallReading: reading.overallReading,
      timestamp: new Date().toISOString(),
    };
  }
}

export class MarketHexagramUseCase {
  constructor(
    private readonly repo: KinhDichRepositoryPort,
    private readonly priceScorePort: PriceScorePort,
    private readonly composerDeps: ReadingComposerDependencies,
  ) {}

  async execute(): Promise<MarketReadingResponse> {
    const code = 'VNINDEX';

    const scores = this.priceScorePort.computeScores(code, 30);
    if (!scores) {
      const stored = this.repo.getLatestReading(code);
      if (!stored) {
        throw new InsufficientDataError(code, 30);
      }
      return {
        hexagram: stored.hexagram_number,
        name: `Que ${stored.hexagram_number}`,
        trend: 'TRUNG TÍNH',
        signal: stored.trading_signal ?? 'GIU',
        confidence: stored.confidence ?? 0.5,
        timestamp: new Date().toISOString(),
      };
    }

    const reading = await composeReading(code, scores, this.composerDeps);
    return {
      hexagram: reading.queChiNh.number,
      name: reading.queChiNh.name,
      trend: reading.queChiNh.trend,
      signal: reading.queChiNh.tradingSignal,
      confidence: reading.queChiNh.confidence,
      timestamp: reading.timestamp,
    };
  }
}

File: apps/kinh-dich-service/pkg/application/usecases.go
// Package application contains the application layer DTOs and use cases.
package application

import (
	"errors"
	"time"

	"github.com/vn-market-intelligence/kinh-dich-service/pkg/domain"
	"github.com/vn-market-intelligence/kinh-dich-service/pkg/module/reading_composer"
)

// ErrNotImplemented is returned when a use case is not yet wired to real implementation.
var ErrNotImplemented = errors.New("kinh-dich ReadingUseCase not implemented — wire reading_composer.ComposeReading before use")

// ErrInsufficientData is returned when there is not enough price data to compute a reading.
var ErrInsufficientData = errors.New("insufficient price data for market reading — requires at least 7 price points")

// ReadingUseCase orchestrates the reading flow.
type ReadingUseCase struct {
	service *domain.ReadingService
}

// NewReadingUseCase creates a new ReadingUseCase.
func NewReadingUseCase(svc *domain.ReadingService) *ReadingUseCase {
	return &ReadingUseCase{
		service: svc,
	}
}

// Execute performs the reading use case.
// Returns ErrNotImplemented until wired to reading_composer.ComposeReading().
// This fail-loud behavior prevents silent fabrication of fake readings.
func (uc *ReadingUseCase) Execute(req ReadingRequest) (*ReadingResponse, error) {
	return nil, ErrNotImplemented
}

// markovAdapter wraps domain.MarkovPort to satisfy reading_composer.MarkovPort.
// This adapter bridges the module-tier MarkovData shape (blending weights) from
// the domain-tier MarkovPort (next-hexagram prediction).
type markovAdapter struct {
	port domain.MarkovPort
}

// GetMarkovData implements reading_composer.MarkovPort.
func (a *markovAdapter) GetMarkovData(hexagramNumber int) (*reading_composer.MarkovData, error) {
	// The domain MarkovPort returns prediction data (nextMostLikely, probability)
	// For now, we return nil since the stub returns nil anyway.
	// In Phase 2, this would convert domain.MarkovData to reading_composer.MarkovData
	// using historicalConfidence derived from prediction accuracy.
	data := a.port.GetMarkovData(hexagramNumber)
	if data == nil {
		return nil, nil
	}
	// Convert domain MarkovData (prediction shape) to module MarkovData (blending shape)
	// Use Probability as both TransitionProb and HistoricalConfidence for now.
	return &reading_composer.MarkovData{
		TransitionProb:       data.Probability,
		HistoricalConfidence: data.Probability,
	}, nil
}

// MarketReading computes a market reading for VNINDEX.
// Returns the current market hexagram, trend, and trading signal.
// Returns ErrInsufficientData if price history is not available.
func (uc *ReadingUseCase) MarketReading() (*MarketReadingResponse, error) {
	const stockCode = "VNINDEX"
	const days = 30

	// Get price scores from PriceScorePort
	priceScorePort := uc.service.GetPriceScore()
	if priceScorePort == nil {
		return nil, ErrInsufficientData
	}

	scores := priceScorePort.ComputeScores(stockCode, days)
	if scores == nil || len(scores) != 6 {
		return nil, ErrInsufficientData
	}

	// Create Markov adapter for reading_composer
	deps := &reading_composer.ReadingComposerDependencies{
		Markov: &markovAdapter{port: uc.service.GetMarkov()},
	}

	// Compose the reading
	reading, err := reading_composer.ComposeReading(stockCode, scores, deps)
	if err != nil {
		return nil, err
	}

	// Map KinhDichReading to MarketReadingResponse
	return &MarketReadingResponse{
		Hexagram:   reading.QueChinh.Number,
		Name:       reading.QueChinh.Name,
		Trend:      reading.QueChinh.Trend,
		Signal:     reading.QueChinh.TradingSignal,
		Confidence: reading.QueChinh.Confidence,
		Timestamp:  time.Now().UTC().Format(time.RFC3339),
	}, nil
}
